python/hail/typecheck/check.py

```diff
@@ -62,21 +117,40 @@ def __init__(self, key_checker, value_checker):
         super(DictChecker, self).__init__()
 
     def check(self, x):
-        passes = isinstance(x, dict)
-        return passes and all(self.kc.check(k) and self.vc.check(v) for k, v in x.iteritems())
+        if not isinstance(x, dict):
+            raise TypecheckFailure
+        x_ = {}
+        kc = self.kc
+        vc = self.vc
+        for k, v in x.items():
+            k_ = kc.check(k)
+            v_ = vc.check(v)
+            x_[k_] = v_
+        return x_
 
     def expects(self):
         return 'dict[%s, %s]' % (self.kc.expects(), self.vc.expects())
 
+    def coerce(self, x):
+        kc = self.kc
+        vc = self.vc
+        return {kc.coerce(k): vc.coerce(v) for k, v in x}
 
-class TupleChecker(TypeChecker):
+
+class SizedTupleChecker(TypeChecker):
     def __init__(self, *elt_checkers):
         self.ec = elt_checkers
         self.n = len(elt_checkers)
-        super(TupleChecker, self).__init__()
+        super(SizedTupleChecker, self).__init__()
 
     def check(self, x):
-        return isinstance(x, tuple) and len(x) == self.n and all(self.ec[i].check(x[i]) for i in range(self.n))
+        if not isinstance(x, tuple):
+            raise TypecheckFailure
+        x_ = []
+        for tc, elt in zip(self.ec, x):
```

Comment:
do you also want to check that x has n elements? length of zip is that of the shorter list.

Reply:
ah yeah, definitely.